src/solver/lu_decomposition.cu

```diff
@@ -4,190 +4,199 @@
 #include <chrono>
 #include <iostream>
 #include <micm/util/cuda_param.hpp>
-#include <vector>
 
-// grouped parameters passing to DecomposeKernel()
-struct DecomposeDevice
-{
-  double* A_;
-  double* L_;
-  double* U_;
-  char* do_aik_;
-  size_t* aik_;
-  char* do_aki_;
-  size_t* aki_;
-  size_t* uii_;
-  std::pair<size_t, size_t>* niLU_;
-  std::pair<size_t, size_t>* uik_nkj_;
-  std::pair<size_t, size_t>* lij_ujk_;
-  std::pair<size_t, size_t>* lki_nkj_;
-  std::pair<size_t, size_t>* lkj_uji_;
-};
 namespace micm
 {
   namespace cuda
   {
-    __global__ void DecomposeKernel(DecomposeDevice* device, size_t n_grids, size_t niLU_size)
+    /// This is the CUDA kernel that performs LU decomposition on the device
+    __global__ void DecomposeKernel(const double* d_A, double* d_L, double* d_U,
+                                    std::pair<size_t, size_t>* d_niLU,
+                                    char* d_do_aik, size_t* d_aik,
+                                    std::pair<size_t, size_t>* d_uik_nkj,
+                                    std::pair<size_t, size_t>* d_lij_ujk,
+                                    char* d_do_aki, size_t* d_aki,
+                                    std::pair<size_t, size_t>* d_lki_nkj,
+                                    std::pair<size_t, size_t>* d_lkj_uji,
+                                    size_t* d_uii,
+                                    size_t niLU_size, size_t ngrids)
```

Comment:
Could the kernel just take as an argument?:
```
LuDecomposeConst& devptr
```

Reply:
The reason I switch to the long list here is that the original implementation had `devptr->xxx` inside the loops and I thought there would be lots of overhead time by doing the deference. But if doing `devptr->xxx` or `devptr.xxx` is as efficient as accessing `xxx` directly, I could try the one argument version as suggested.

Reply:
A compromise could be passing in a `LuDecomposeConst&` and saving all of the arguments here as local variables. It increases readability while preventing the performance hit that you're concerned about since all data access would then be stack variables

Reply:
Thanks @K20shores . I just did what you suggest (hopefully I understand it correctly) and it worked.

One thing I need to mention: using `LuDecomposeConst&` passes the compilation but the execution of this CUDA test will hang. Switching to passing by value (i.e., `LuDecomposeConst`) fixes the runtime problem but I do not understand why.

Reply:
A potential source of the intermittent failures with pointers and references could be that there is a default constructor that doesn't define `CudaLuDecomposition::devptr_` [here](https://github.com/NCAR/micm/blob/2686dc09f057eb32d9e28fdb96a9cc1cc4ded8a8/include/micm/solver/cuda_lu_decomposition.hpp#L22-L23) but `CudaLuDecomposition::Decompose` assumes that it is in a valid state [here](https://github.com/NCAR/micm/blob/2686dc09f057eb32d9e28fdb96a9cc1cc4ded8a8/include/micm/solver/cuda_lu_decomposition.hpp#L98). If the default constructor is needed, maybe there could be some check that `devptr_` is in a valid state?

Reply:
Oh, that could be the case. But do we really need the default constructor for MICM? Or it has to exist?

Reply:
@sjsprecious it doesn't necessarily have to exist. You can delete it or make it private and see what happens.